lib/src/klio/transforms/core.py

```diff
@@ -80,6 +84,8 @@ def get(cls):
         with cls._lock:
             if cls._config is None:
                 cls._config = cls._load_config_from_file()
+                if cls._config.job_config.use_main_session_config is True:
```

Comment:
Is there a way to not need user config at all, and just do something (either at the top of the module or within this class):

```py
from apache_beam import __version__ as beam_version

# snip

use_main_session = False  # maybe this is a class var? 
major, minor, patch = beam_version.split(".") 
if int(major) == 2 and int(minor) >= 31:
    use_main_session = True
```

Reply:
I think we could definitely do that eventually, maybe in `21.10.0`, but for now at least I want to keep this strictly opt-in just in case.

Reply:
Could you tell me a bit more as to why you want to keep this as strictly opt-in? Or maybe another way to ask it: why wouldn't users want to use_main_session if they're using a version that has the fixed bug?

What I'm concerned about is adding a new configuration setting, particularly if the plans are to deprecated in the following release.

Is it the case that we're not confident it will work/worried the session may eventually be lost? and therefore we want to have an "out" to let users toggle it off?

Reply:
Yes, my concern is that we never definitively proved that the issues we ran into with using main session last time were in fact caused by the bug in Beam.  It obviously seems very likely the cause, but being unable to reliably reproduce the issue in testing also means I was unable to verify that the issue no longer exists.

Giving it some more thought though, it might just be better to enable it without having a config setting, and if someone actually runs into this issue again, we'd just push out a hot-fix reverting main session again, and whoever ran into it would just rollback to the previous release.

I'm hopeful that we can get more extensive testing done in a release candidate, so if we do have to flip back it happens before we push out the final release.

Lastly, I just checked and Beam is currently in release candidate for 2.33, which means if we time things right we can push this out without having to check Beam version, since the earliest version we'd support would be 2.31

Reply:
Ok cool. I know we decided long ago that we'd only support the latest 3 versions of beam, but do users update their beam version along-side their Klio version? If you don't want to do the version check (which is fine), could we specify the minimum beam version instead?

Reply:
> do users update their beam version along-side their Klio version?

I think there is a mix of some users not having a beam version specified in their `requirements.txt` and simply default to whatever Klio specifies (or what may already be included in their docker image), and other users having beam pinned to a specific version.

Either way, I think it's reasonable to recommend to users that they are always using a Beam version supported by the Klio version they're using.

> could we specify the minimum beam version instead?

We already have a min version in `setup.py` for `>=2.21.0` (which is very out of date anyway), so I can update that and update the tested versions in `tox.ini`.

It probably also makes sense to leave this unmerged until Beam 2.33 is released, unless we're ok temporarily only supporting 2 released beam versions, or targeting `2.33rc1` as the latest release

Reply:
ok sounds good to me! ty!